crates/bevy_ecs/src/query/filter.rs

```diff
@@ -512,6 +512,183 @@ macro_rules! impl_tuple_query_filter {
+/// #
+/// # #[derive(Component)]
+/// # struct MyComponent;
+/// #
+/// fn print_mutated_component_system(
+///     query: Query<Entity, (Changed<MyComponent>, Not<Added<MyComponent>>)>
+/// ) {
+///     for entity in &query {
+///         println!("Entity {:?} had MyComponent changed, but was not just added", entity);
+///     }
+/// }
+/// # bevy_ecs::system::assert_is_system(print_mutated_component_system);
+/// ```
+pub struct Not<T>(PhantomData<T>);
+
+/// SAFETY:
+///
+/// For purely archetypal filters, this inverts the access in `update_component_access` as well as in
+/// `matches_component_set`, ensuring that they match.
+///
+/// For purely non-archetypal filters, this maintains the same read accesses as the original filter in
+/// both instances.
+
+unsafe impl<T> WorldQuery for Not<T>
+where
+    T: QueryFilter + InvertibleFilter,
+{
+    type Item<'a> = T::Item<'a>;
+
+    type Fetch<'a> = T::Fetch<'a>;
+
+    type State = T::State;
+
+    fn shrink<'wlong: 'wshort, 'wshort>(item: Self::Item<'wlong>) -> Self::Item<'wshort> {
+        T::shrink(item)
+    }
+
+    #[inline]
+    unsafe fn init_fetch<'w>(
+        world: UnsafeWorldCell<'w>,
+        state: &Self::State,
+        last_run: Tick,
+        this_run: Tick,
+    ) -> Self::Fetch<'w> {
+        T::init_fetch(world, state, last_run, this_run)
+    }
+
+    const IS_DENSE: bool = T::IS_DENSE;
+
+    #[inline]
+    unsafe fn set_archetype<'w>(
+        fetch: &mut Self::Fetch<'w>,
+        state: &Self::State,
+        archetype: &'w Archetype,
+        table: &'w Table,
+    ) {
+        T::set_archetype(fetch, state, archetype, table);
+    }
+
+    #[inline]
+    unsafe fn set_table<'w>(fetch: &mut Self::Fetch<'w>, state: &Self::State, table: &'w Table) {
+        T::set_table(fetch, state, table);
+    }
+
+    #[inline(always)]
+    unsafe fn fetch<'w>(
+        fetch: &mut Self::Fetch<'w>,
+        entity: Entity,
+        table_row: TableRow,
+    ) -> Self::Item<'w> {
+        T::fetch(fetch, entity, table_row)
+    }
+
+    fn update_component_access(state: &Self::State, access: &mut FilteredAccess<ComponentId>) {
+        T::update_component_access(state, access);
+
+        for filter in &mut access.filter_sets {
+            std::mem::swap(&mut filter.with, &mut filter.without);
+        }
+    }
+
+    fn update_archetype_component_access(
+        state: &Self::State,
+        archetype: &Archetype,
+        access: &mut Access<ArchetypeComponentId>,
+    ) {
+        T::update_archetype_component_access(state, archetype, access);
+    }
+
+    fn init_state(world: &mut World) -> Self::State {
+        T::init_state(world)
+    }
+
+    fn matches_component_set(
+        state: &Self::State,
+        set_contains_id: &impl Fn(ComponentId) -> bool,
+    ) -> bool {
+        if T::IS_ARCHETYPAL {
+            !T::matches_component_set(state, set_contains_id)
+        } else {
+            T::matches_component_set(state, set_contains_id)
```

Comment:
I find this a bit questionable since it means that `Or<(Added<A>, Not<Added<A>>)` won't match every entity, but only those that have `A`. This is also the fundamental problem you face when filter being negated is a tuple mixing `With`/`Without` and `Added`/`Changed`. In one case you want to negate the archetypes being matched, in the other you want to negate the `filter_fetch`, but in case of tuples you need a mix of them.

Reply:
I don't think that it makes sense for `Not<Changed<A>>` to also match entities that do not have `A` at all. It's not the correct logical negation, but is probably the expected result when filtering on `Not<Changed<A>>`.

Reply:
While I agree that the expectation is different, it also means that the operation that `Not` is doing is different for different types, and this can break all sorts of expectations in the code.

Here's yet another test case that fails:

```rs
    #[test]
    fn query_filter_not_broken() {
        use crate::system::Query;
        fn system(
            q1: Query<(Entity, &mut A), Not<Changed<B>>>,
            q2: Query<(Entity, &mut A), With<B>>,
        ) {
            for (e1, _) in &q1 {
                for (e2, _) in &q2 {
                    assert_ne!(e1, e2);
                }
            }
        }

        let mut world = World::default();
        world.spawn((A(0), B(0)));
        let system_id = world.register_system(system);
        _ = world.run_system(system_id);
        _ = world.run_system(system_id);
    }
```

The problem is that you always invert the component access, but then when setting which component sets are matched by `Not<Changed<B>>` you don't invert them, hence declaring that `Not<Changed<B>>` accesses only archetypes without `B` (making it disjoint from the second query) but in the end accessing only those that have `B` (making it conflicting with the second query).

---

If I had to generalize the operations `Not` actually does:

- there is an "applicable components set" on which a filter can be applied on. Every archetype not matching it is always excluded by it. For `With`/`Without` this set contains every possible component set, but for `Added<T>`/`Changed<T>` this contains only component sets containing `T`s. This is not inverted by `Not`, and should be kept as is.
- there is a "filtered components set" which is a subset of the applicable components set that the filter can reduce to. For example `With<T>`/`Without<T>` can reduce the applicable components set to only those containing `T` or not. `Not` inverts this filter.
- there is a "fetch filter" that filters individual entities matching the filtered components set. For example `Added<T>`/`Changed<T>` apply filters at this point. `Not` also inverts this filter.

The problem with this approach is that it adds yet another level of component access/filtering, complicating `WorldQuery` even more. But without it I don't think you can reasonably support inverting both `With`/`Without` and `Changed`/`Added` without bugs.

Reply:
I think `Not<Changed<T>>` should include entities which do not contain a `T`. I think the problem here is the wording between `Not Changed T` and an `Unchanged T`. `Unchanged T` implies the existence of `T`, whereas `Not Changed T` is strictly the inverse of `Changed T`.

A resolution to this problem could be to define:

```rust
pub type Unchanged<T> = (With<T>, Not<Changed<T>>);

pub type Existing<T> = (With<T>, Not<Added<T>>);
```

With some documentation explaining why it's not as simple as `Not<Changed<T>>`. For something as fundamental as a `Not` operation, it should be identical to the laws of boolean algebra.